fusor/mig-controller
# Azure BSL support

Adds support for Azure as a BackupStorageLocation / Migration Registry storage provider.

## Test Instructions

First, ensure that you're deploying other mig components with recent version of mig-operator. This is necessary to ensure that the AZURE_CREDENTIALS_FILE env var is set. https://github.com/fusor/mig-operator/blob/master/roles/migrationcontroller/templates/velero.yml.j2#L60-L61

```
# Create Azure storage Resource Group

AZURE_RESOURCE_GROUP=Velero_Backups
az group create -n $AZURE_RESOURCE_GROUP --location CentralUS
```

```
# Create Azure Storage Account

az storage account create \
     --name $AZURE_STORAGE_ACCOUNT_ID \
     --resource-group $AZURE_RESOURCE_GROUP \
     --sku Standard_GRS \
     --encryption-services blob \
     --https-only true \
     --kind BlobStorage \
     --access-tier Hot
```

```
# Create Azure Blob Storage Container

BLOB_CONTAINER=velero
az storage container create -n $BLOB_CONTAINER --public-access off --account-name $AZURE_STORAGE_ACCOUNT_ID
```

```
# Create Service Principal for Velero to act as

AZURE_SUBSCRIPTION_ID=`az account list --query '[?isDefault].id' -o tsv`
AZURE_TENANT_ID=`az account list --query '[?isDefault].tenantId' -o tsv`
AZURE_CLIENT_SECRET=`az ad sp create-for-rbac --name "velero" --role "Contributor" --query 'password' -o tsv`
AZURE_CLIENT_ID=`az ad sp list --display-name "velero" --query '[0].appId' -o tsv`
```

```
# Dump secret credentials to file that we'll feed into creds secret

cat << EOF  > ./credentials-velero
AZURE_SUBSCRIPTION_ID=${AZURE_SUBSCRIPTION_ID}
AZURE_TENANT_ID=${AZURE_TENANT_ID}
AZURE_CLIENT_ID=${AZURE_CLIENT_ID}
AZURE_CLIENT_SECRET=${AZURE_CLIENT_SECRET}
AZURE_RESOURCE_GROUP=${AZURE_RESOURCE_GROUP}
AZURE_CLOUD_NAME=AzurePublicCloud
EOF
```

```
# Load b64 encoded contents of credentials into mig-storage-creds

cat << EOF  > ./mig-storage-creds-azure.yaml
---
apiVersion: v1
kind: Secret
metadata:
  namespace: openshift-migration
  name: migstorage-creds
type: Opaque
data:
  azure-credentials: $(base64 credentials-velero -w 0)

EOF


oc create -f mig-storage-creds-azure.yaml
```

```
# Create migstorage configured against Azure Blob Storage

cat << EOF  > ./mig-storage-azure.yaml
---
apiVersion: migration.openshift.io/v1alpha1
kind: MigStorage
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: migstorage-sample
  namespace: openshift-migration
spec:
  backupStorageProvider: azure
  volumeSnapshotProvider: azure

  backupStorageConfig:
    azureStorageAccount: ${AZURE_STORAGE_ACCOUNT_ID}
    azureStorageContainer: ${BLOB_CONTAINER}
    azureResourceGroup: ${AZURE_RESOURCE_GROUP}

    credsSecretRef:
      namespace: openshift-migration
      name: migstorage-creds

  volumeSnapshotConfig:
    azureResourceGroup: ${AZURE_RESOURCE_GROUP}
    azureApiTimeout: 30s
    credsSecretRef:
      namespace: openshift-migration
      name: migstorage-creds

EOF

oc create -f mig-storage-azure.yaml
```
